### PR TITLE
Fix sandbox only API

### PR DESCRIPTION
**Pull request by AmaliMatharaarachchi**
### Issues

Fixes https://github.com/wso2/product-microgateway/issues/2542

### Automation tests
 - Unit tests added: No
 - Integration tests added: Yes. modified existing test case to catch this issue

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

**Review by praminda — COMMENTED** [review]
@AmaliMatharaarachchi Error in the issue sounds like it doesn't consider the sandbox endpoint even if you have given it via the extension. I'm wondering how this fix affects that scenario?
Basically if you have openapi server definition (Ep A) + sandbox ep extension (Ep B) to override server definition for sandbox tokens behavior should be,
1. invoking with prod token -> response of Ep A
2. invoking with sand token -> response of Ep B
isn't it? or did i misunderstood the case here?

File: adapter/internal/oasparser/model/mgw_swagger.go [hunk 1]
```diff
@@ -380,18 +380,24 @@ func (swagger *MgwSwagger) SetXWso2Extensions() error {
...
-	sandboxEndpointErr := swagger.setXWso2SandboxEndpoint()
+	apiLevelSandEPFound, sandboxEndpointErr := swagger.setXWso2SandboxEndpoint()
 	if sandboxEndpointErr != nil {
 		logger.LoggerOasparser.Error("Error while adding x-wso2-sandbox-endpoints. ", sandboxEndpointErr)
 		return sandboxEndpointErr
 	}
 
+	// to remove swagger server/host urls being added when x-wso2-sandbox-endpoints is given
+	if !apiLevelProdEPFound && apiLevelSandEPFound && swagger.productionEndpoints != nil &&
+		len(swagger.productionEndpoints.Endpoints) > 0 {
```
[the remark this answers is not in the document]
Comment (renuka-fernando) [reply adapter/internal/oasparser/model/mgw_swagger.go:397]: Here `swagger.productionEndpoints != nil` is there ryt? if `Endpoints` is nil len will return 0

[the remark this answers is not in the document]
Comment (praminda) [reply adapter/internal/oasparser/model/mgw_swagger.go:397]: ah wait i missed the two variables being checked here. I saw we were checking both `nil` and `len > 0` for `swagger.productionEndpoints`

**Comment by AmaliMatharaarachchi** [comment]
> @AmaliMatharaarachchi Error in the issue sounds like it doesn't consider the sandbox endpoint even if you have given it via the extension. I'm wondering how this fix affects that scenario? Basically if you have openapi server definition (Ep A) + sandbox ep extension (Ep B) to override server definition for sandbox tokens, the behavior should be,
> 
> 1. invoking with prod token -> response of Ep A
> 2. invoking with sand token -> response of Ep B
>    isn't it? or did i misunderstood the case here?
@praminda,
This issue happens due to the below previous behaviour,
when, x-wso2-production-endpoints is not given, swagger servers object/host fields will generate the production endpoints (in the issue, petstore api have host 'petstore.io', so now prod endpoint is 'petstore.io'). regardless of having x-wso2-sandbox-endpoints, this production endpoint will be retrieved from this(in the issue, only x-wso2-sandbox is present in swagger, so production endpoint will be populated from swagger servers, sandbox will be populated from extension). now, prod and sand are both not empty however their basepaths are different. hence, prod endpoints will be assigned, sand will be ignored.

**Review by pubudu538 — APPROVED** [review]

**Comment by praminda** [comment]
> > @AmaliMatharaarachchi Error in the issue sounds like it doesn't consider the sandbox endpoint even if you have given it via the extension. I'm wondering how this fix affects that scenario? Basically if you have openapi server definition (Ep A) + sandbox ep extension (Ep B) to override server definition for sandbox tokens, the behavior should be,
> > 
> > 1. invoking with prod token -> response of Ep A
> > 2. invoking with sand token -> response of Ep B
> >    isn't it? or did i misunderstood the case here?
> 
> @praminda, This issue happens due to the below previous behaviour, when, x-wso2-production-endpoints is not given, swagger servers object/host fields will generate the production endpoints (in the issue, petstore api have host 'petstore.io', so now prod endpoint is 'petstore.io'). regardless of having x-wso2-sandbox-endpoints, this production endpoint will be retrieved from this(in the issue, only x-wso2-sandbox is present in swagger, so production endpoint will be populated from swagger servers, sandbox will be populated from extension). now, prod and sand are both not empty however their basepaths are different. hence, prod endpoints will be assigned, sand will be ignored.
> 
> This fix is to override swagger servers endpoint population, when either production or sandbox is provided by the extension. it will solve the above issue.

Is this the expected behavior when servers object's basepath and extension's basepath is equal?

**Comment by AmaliMatharaarachchi** [comment]
@praminda IMHO, it's not the expected behaviour even if basepaths of servers object and extension are matched. If production-endpoints or sandbox-endpoints extension is there, we should not read the servers object.

**Review by praminda — APPROVED** [review]

**Comment by renuka-fernando** [comment]
Integration tests have passed in the PR check workflow
only the stage "Build multi platform Ubuntu images" is failing, and this is fixed in #2549
Hense merging the PR

**Comment by choreo-cicd** [comment]
[failed] dev-deployment : 20211215.32

**Comment by choreo-cicd** [comment]
[succeeded] dev-deployment : 20211215.43

**Pull request merged**